### PR TITLE
Issue 440 - Loads the newest test engine found

### PR DESCRIPTION
**Pull request by rprouse**
Fix for #440

I am pushing this early for discussion. It is based on the API that @CharliePoole and I discussed to load the engine,

```C#
public static ITestEngine CreateInstance(bool privateCopy = false);
public static ITestEngine CreateInstance(Version minVersion, bool privateCopy = false);
```

These will attempt to load the newest test engine that can be found that is `>= minVersion`. If `minVersion` is not specified, it will load any engine greater than or equal to 3.0.

If `privateCopy` is true, we will only look in the Application Base directory for the engine. This is to allow third party runners to only use the copy of the engine that they ship and have tested with if they wish.

If `privateCopy` is false, the default, the current search order is,

1. The application base directory.
2. An engine installed to the current user's account as found in the 32 bit registry key `HKCU\SOFTWARE\Nunit.org\Engine\{version}`.
3. An engine installed to the local machine as found in the 32 bit registry key `HKLM\SOFTWARE\Nunit.org\Engine\{version}`.

If multiple engines are found, the newest engine is returned.

~~I have not yet modified the install to add the registry key when installing the engine.~~

So far, this is a fairly simplistic search and not terribly useful on Linux. Do we want to search other directories? If so, how will they be located? Some ideas that have been discussed are;

- Installed in the GAC (personally, I don't want to see our engine installed in the GAC)
- A directory specified in the local nunit.settings file (Charlie mentioned a problem in that the engine is currently required to read the settings file)
- Another config file?
- Search the PATH?
- ~~We do not currently support multiple installs of the engine. We want users to use the latest engine. Should that be rethought?~~
- Should we create a Linux installer that registers the engine in the Mono registry?

### CODE REVIEW

**Comment by rprouse** [comment]
If an engine that does not meet the min/max version is found, an `NUnitEngineNotFoundException` is thrown.

**Comment by remcomulder** [comment]
Thanks for opening a discussion on this and for being so inclusive about it.  It's great to be able to have a conversation about how this API works before we're all integrated with it and supporting our users as they test with it.

Cross-versioning is a huge problem for test runner authors such as myself and historically these sorts of problems tend to be well represented in user support requests, even well after all integration has been stabilised .. so I take a great interest in how this is implemented :)

Before giving my own 2c on this, I want to say that I don't want to come across as too negative, especially when there has already been work done and time invested in trying to solve this problem.  With this said, I have to say that I wonder if this is, in fact, an elegant solution to the wrong problem.

Assuming that a test framework is packaged and deployed in a consistent way with its execution engine (for example, using the same rules in Nuget), resolving an execution engine is not a difficult problem for a test runner to solve, and can usually be done in a few short lines of code.  The major problems with cross-versioning tend to revolve around changes in interaction between the runner and the engine (i.e. API changes).

By creating a library that encapsulates the problem of resolving a test framework, this encourages the formation of multiple rules (i.e. search paths) detailing where execution engine assemblies are loaded from, and serves to 'hide' these rules from the runner attempting to resolve the assemblies.  IMHO, this adds little value and can make products more difficult to support, as it then becomes harder to know for certain where a user's execution engine assemblies are being loaded from.

A good example of the pain involved in dynamic assembly loading is the CLR itself.  Because the CLR has so many rules that describe where assemblies can be loaded from (i.e. application base directory, GAC, etc), sometimes it can be difficult to be certain where they are coming from.  In certain cases, the loading logic can make it impossible for an application to exert control over the source of the assemblies being loaded, resulting strange defects that can be difficult to reproduce outside of the user's environment.  I realise the rules being suggested here are very simple, but it doesn't take much for this indirection to take a toll on development and product support.

The simplest solution to the assembly resolution problem is to avoid it entirely.  Place the NUnit runner code inside the framework assembly containing the test attributes.  In this way, the runner can just rely on the assembly already referenced by the user's code.  This ensures the correct version of the runner always goes with the correct version of the attributes, and no indirection is required.  This is the approach taken by MSpec, and it works really well.  After integrating with 6 different test frameworks and supporting this integration for 5 years, MSpec has given me the least trouble of all frameworks almost entirely because of this design.

Thanks!

Remco Mulder
NCrunch Developer

**Comment by CharliePoole** [comment]
Hi Remco,

"Place the NUnit runner code inside the framework assembly..."

I tried to respond to this in an earlier post, but I think I failed to find
the right words. Let me try again.

"Runner" is a highly overloaded word, with interpretations ranging from
NCrunch itself to the code in nunit.core that interprets attributes. I'm
imagining that you mean the latter, the lowest level "runner" of them all.
This is the code that knows, for example, what a TestAttribute means and
causes a test to be created using the method it decorates.

[In the very earliest days of NUnit 2.0, this logic was included in the
framework assembly. We thought it was a good idea to separate it into the
nunit.core assembly and so we did. It turns out to create lots of
difficulties, especially for third-party runners, but even for ourselves.
If new attributes are introduced, you have to have a new core that
understands them and it's necessary to always use the latest core in order
to interpret the latest features. It just pushes the versioning issue back
one level of indirection.]

With NUnit 3.0, the core that interprets attributes, finds tests and
executes them is back in the framework. There is no longer a separate
'core'. The user code references the framework, so the proper code is
always loaded to interpret the tests. If I understand correctly and that's
what you are wanting us to do, it's already done.

The engine is a much higher-level interface. For a broad look at what the
engine is (although a little out of date) see
https://github.com/nunit/dev/wiki/Architectural-Overview

Here's a list of some things the engine does directly:

   * Enables running tests in various places: AppDomains, Processes,
separate machines, devices.
   * Separates the runner from the type of framework. For example, the
nunit console can also run tests built against V2 of NUnit.
   * Runs multiple assemblies and combines the results
   * Creates the top-level element in the XML output
   * Reformats XML output in various formats, including user-specified
transforms and NUnit v2 format output
   * Has an addin-based design allowing addition of new facilities.

Here is some stuff it does not do itself, but asks the framework to do:

   * Discovering tests
   * Executing tests
   * Filtering tests
   * Producing XML output of test results
   * Associating text output with the test results


The engine and framework are connected via a driver. The V3 driver is built
in and the V2 driver is delivered as an addin. Somebody could build a
driver for xUnit, and the NUnit console and (future) gui would run the
tests.

You don't have to use the engine, but I'm hoping that you and other authors
will. It gives access to the features implemented in the engine, which
should benefit the users. For example, when we add the ability to run tests
on a separate machine, that will be an engine feature. You could (most
likely) access that feature by adding a name / value pair to the test
package.

OTOH, you could use the framework directly, writing code similar to (or
copying) the V3 framework driver that's part of the engine. You would then
automatically gain access to any features added in the framework, because
the "runner" for those features is a part of the framework. You would not
get any higher-level features we might add and you couldn't use any addins
(which also deal with higher-level features). For some runners, this may be
the best choice, and we can discuss that interface here as well.

Even if you think you may bypass the engine, I hope you will still
contribute to the discussion and work through the pros and cons. One area
we have not yet touched on is the actual content of the XML that is
returned to you from a test run and as progress notices during the run.
That comes from the framework and is passed directly to you by the engine,
so it's the same in either case. If that XML doesn't contain what you need,
then you have a big problem, so it's important we get it right.

Hopefully, this puts the engine in perspective for the moment. Frankly, if
you are using an outdated engine, I don't think it's that big a deal. It
will still work. We expect an older engine to work with a newer framework,
because the dictionary of parameters is simply passed through. My guess is
that 90% of improvments to NUnit will be to the framework alone, with the
remaining 10% divided among the engine and our runners.

Charlie

**Comment by remcomulder** [comment]
Thanks Charlie.  I'm really glad now that I had completely misunderstood the architecture here :)

I'll need to take a deeper look to be certain, but I will say that most likely integrating directly with the Framework seems like the best approach for NCrunch.  Many of the features provided by the Engine overlap with NCrunch's own features, and trying to navigate an integration path to get the best of both worlds will likely end up with a complex mess for everyone.

I did try to integrate with higher level runners/engines while integrating with Gallio/Xunit and MSTest ... in each case I ended up needing to integrate at a deeper level.  MSTest was actually so bad that I had to avoid integrating entirely and just wrote an emulator instead.

The content of the XML itself will definitely be critical in all of this.  NCrunch itself doesn't consume much from framework execution runs, but it does tend to be very picky about needing to control the environment around the execution.

**Comment by CharliePoole** [comment]
Does NCrunch arrange to run the tests in a process under the correct
framework?  That's an engine function I forgot to list.

Charlie

**Comment by CharliePoole** [comment]
Here's the V3 FrameworkDriver from our engine:
https://github.com/nunit/nunit/blob/master/src/NUnitEngine/nunit.engine/Drivers/NUnit3FrameworkDriver.cs

Here's the corresponding code in the framework:
https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Api/FrameworkController.cs

You'll recognize this as inspired by xUnit. We expect to keep this pretty
much constant as we do new releases.

Charlie

**Comment by remcomulder** [comment]
Yes – it also deals with higher level environment integration with a number of other tools and frameworks, including Windows Store sandboxing, MS Fakes, JustMock, TypeMock, etc.  Basically it takes responsibility for building the entire environment around the test execution run.  If the NUnit Engine takes responsibility for launching any .exe or building its own application domain, then there’s no way I’ll be able to work with the Engine and I’ll definitely need to integrate directly with the Framework.

**Comment by remcomulder** [comment]
Thanks.  It looks like it’s best if I go straight to the controller … I think I did something similar with Xunit.

 

Can tests only be identified for execution using the XML filter?  If there’s a way to execute them directly by passing in a list of IDs or fixture type names, this would perform much better.  Otherwise I expect the Framework would need to enumerate the entire assembly to apply the filter.

 

Do also let me know if there’s a better place to discuss this.  I realise we’re getting a bit off topic from Rob’s commit :)

**Comment by CharliePoole** [comment]
I've been responding by email and didn't realize we were not on the mail list. I'll go there.

**Comment by rprouse** [comment]
I have pushed new changes. My initial PR comment has been updated to reflect the changes.

1. The install now registers the engine.
2. I changed the search order so it looks in the current directory first.
3. I now check the Wow64 registry node for the engine also.

**Comment by CharliePoole** [comment]
Hmmm.... If we register the engine in the install, then it will only be registered for users who run the install. We could register it on first run, but that creates a slight chicken and egg problem. Have to think about that.

**Comment by rprouse** [comment]
I think this is ready to merge. We haven't received any additional feedback on the search locations or order and they are very easy to change. I tested by installing the engine using the installer and then removing it and all addins from the build directory. I then ran tests using the console in the build directory and it used the installed engine.

The change to mono.addins is because the default initialization uses the entry assembly. This became a problem causing failures when the engine was being loaded from a different location (the install location) and might result in addins being loaded from the runner directory rather than the engine directory. If we use addins outside of the engine, those should be initialized as they were before.

I agree that we need to think about other ways of registering the engine. Personally, I think we should wait and see how it pans out and we understand how this will be used.

**Comment by CharliePoole** [comment]
The code looks good. Let's re-review the concept before merging. Some questions:

1. How do we deal with problem of users who don't use the install and therefore don't have an entry in the registry? Let's assume you just unzip the files somewhere (you're a developer) and you want NUnit to be aware of the engine location.

2. New installs will overwrite old ones in the engine. Is that what we want? Do we want to allow for multiple versions to be recorded in the registry - e.g.: for use by old software?

3. Is the privateCopy setting needed? The downside is that users can't defeat the setting except by copying an assembly to the local directory. What happens if we simply use any engine in the path that meets the requirements and don't look further unless we haven't found one.

4. Let's think about two scenarios:
  * User wants to upgrade a runner to use a newly installed engine
  * User wants some runner to keep using the previous engine, even though a new one is installed

**Comment by rprouse** [comment]
I'm not sure that I have answers to all of these questions, that is why I was suggesting that we merge and make additions if and when we receive feedback. @remcomulder has already stated that this feature is probably not useful for him, so I think the major consumers will likely be our runners.

>How do we deal with problem of users who don't use the install and therefore don't have an entry in the registry? Let's assume you just unzip the files somewhere (you're a developer) and you want NUnit to be aware of the engine location.

I'm not sure how we handle this. MbUnit has a small settings program that handled the registrations for Zip installs, but in practice, none of the developers in my company bothered to run it and wondered why things were broken.

>New installs will overwrite old ones in the engine. Is that what we want? Do we want to allow for multiple versions to be recorded in the registry - e.g.: for use by old software?

Personally, I would prefer that newer installs of the engine overwrite older installs, but that I can install multiple versions of the framework.

> Is the privateCopy setting needed? The downside is that users can't defeat the setting except by copying an assembly to the local directory. What happens if we simply use any engine in the path that meets the requirements and don't look further unless we haven't found one.

I think it is needed. It allows developers of runners to specify that they never want to use a version other than the one they shipped. If we return the first engine that meets the requirements, then it defeats the purpose of this feature. We could switch it to privateCopy to mean return the engine from the local directory if it meets specifications, only search elsewhere if it does not meet the requirements.

> User wants to upgrade a runner to use a newly installed engine

> User wants some runner to keep using the previous engine, even though a new one is installed

I don't think that should be the user's choice, it should be the developer's choice. That is part of the reason that I think this issue might be a solution looking for a problem. It sounds great on paper, but if we give the user's the ability to switch around engines, then we could be creating a support nightmare. 

Developers investigating issues will have a difficult time knowing which engine users are running or where it came from. It will make it very difficult to determine if problems are actual bugs or incompatibilities between engines. Personally, if I was developing a runner, I would skip that headache and ship a known good engine and use `privateCopy = true`.

I've always like the idea of this feature and the flexibility, but when I put my pessimist hat on, it doesn't pass muster :smile:

I'm not saying we should abandon it, just that we should see if and how it is used before we kill ourselves over it.

**Comment by CharliePoole** [comment]
Hi  Rob,

I'll try to organize this into separate points...

1. Users who don't install: I think we can leave this for now. One way I have dealt with it in the past is to have a program write an entry in the registry on first execution if it doesn't already find one there. That brings in some issues of permissions, so let's wait on this one.

2. " I would prefer that newer installs of the engine overwrite older installs, but that I can install multiple versions of the framework." Can you explain that a bit more? If old ones are overwritten, then how can you install multiple versions? Should we have a key like v3.0, v3.2, etc?

3. Private copy: I was suggesting something a bit different. The presence of an engine assembly would mean privateCopy, without ever needing to specify it in the call. The search order would be. If a private copy is found meeting the min and max requirements, use it. Otherwise continue the search. IOW, I think we are specifying we want a private copy twice: once in the call and once by virtue of the fact that it's present.

If we allow the privateCopy value in the call, then I am suspecting that all 3rd party runners would use it, depriving users of  the ability to upgrade. It also encourages them to modify the engine for their own purposes, relying on special features they add in a proprietary manner.

4. User wants to upgrade: This should probably be first, because it drives a lot of the rest for me. Because of past problems with TD.Net and R#, I wanted to give users the choice of forcing the 3rd party runner to use a new engine we have released. With the current API, they only have that choice if the vendor has allowed it. I'll allow that this might be OK. If the vendor isn't sufficiently responsive in terms of using a new engine, then pressure will be brought on them to change how they use the API and allow users to do that. 

Still, it seems to me that vendors knowing their code may be used with any future build of the engine would force them to write it in a safe way.

Note that those who use the framework directly will already have to work this way. It's _only_ possible for the framework API to use the framework that the user has referenced.

5. New point: If most vendors end up _not_ using the engine, then maybe we should reduce the API to those features we want to use. I think that means minimum version only and no private copy.

*******

I agree some of these don't have to be resolved before the beta. I'd like to settle on our best guess as to the engine API and the form of registry entries if possible. In particular, for the registry entries, it will be annoying if we have to deal with changing it in the install for the second beta.

Charlie

**Comment by rprouse** [comment]
> Users who don't install: I think we can leave this for now. One way I have dealt with it in the past is to have a program write an entry in the registry on first execution if it doesn't already find one there. That brings in some issues of permissions, so let's wait on this one.

Agreed

> "I would prefer that newer installs of the engine overwrite older installs, but that I can install multiple versions of the framework." Can you explain that a bit more? If old ones are overwritten, then how can you install multiple versions? Should we have a key like v3.0, v3.2, etc?

If we say that engines will continue to be backwards compatible for the foreseeable future, then I don't see a need to have multiple versions of the engine installed. If we are not sure that will be the case, then we could say that minor revisions (third digit) overwrite themselves, but major (first or second digit) install side by side. We could also make that cut when we introduce potentially breaking changes. Either way, it might make sense to include the version in the registry key. **I will make that change and we can decide on the install.** Sound good?

> Private copy: I was suggesting something a bit different. The presence of an engine assembly would mean privateCopy, without ever needing to specify it in the call. The search order would be. If a private copy is found meeting the min and max requirements, use it. Otherwise continue the search. IOW, I think we are specifying we want a private copy twice: once in the call and once by virtue of the fact that it's present.

> If we allow the privateCopy value in the call, then I am suspecting that all 3rd party runners would use it, depriving users of the ability to upgrade. It also encourages them to modify the engine for their own purposes, relying on special features they add in a proprietary manner.

I would disagree with the presence of the engine meaning private copy. If I was a third party runner, I would include the engine with my install in case the user doesn't have it installed, but if they have a newer engine installed, I would use that (`privateCopy = false`).

`privateCopy = true` would then be available for people that write runners but do not want to support using newer runners. I think it should be their choice, not their users because it is them, not us, that will need to support their software.

>User wants to upgrade: This should probably be first, because it drives a lot of the rest for me. Because of past problems with TD.Net and R#, I wanted to give users the choice of forcing the 3rd party runner to use a new engine we have released. With the current API, they only have that choice if the vendor has allowed it. I'll allow that this might be OK. If the vendor isn't sufficiently responsive in terms of using a new engine, then pressure will be brought on them to change how they use the API and allow users to do that. 

>Still, it seems to me that vendors knowing their code may be used with any future build of the engine would force them to write it in a safe way.

>Note that those who use the framework directly will already have to work this way. It's only possible for the framework API to use the framework that the user has referenced.

I see your point, but I still stand by my previous points. I think we should encourage, but not force, developers that write runners to use the API correctly and to set `privateCopy = false`. If we do remove the privateCopy, I expect almost all runners will ship the engine with their runner effectively making it private copy. In that case, they will have the same incentive to use the API incorrectly. In both cases, if the user wants to update the engine, they can drop it in the runner directory and hope that it works.

I am not strongly attached to this though, just making my point, so make the call @charliepoole and I will implement it accordingly.

>New point: If most vendors end up not using the engine, then maybe we should reduce the API to those features we want to use. I think that means minimum version only and no private copy.

I also wonder if that will be the case, but I think we should make it as flexible as possible to encourage them to use it. I will leave `privateCopy` up to you as per above, but I think we can probably remove `maxVersion` since we will never know at what point compatibility will be broken and once that happens, you would likely update the runner and set `minVersion` instead.

**Comment by CharliePoole** [comment]
I think we are in agreement. Point by point, then...

On Sun, Mar 8, 2015 at 10:17 AM, Rob Prouse <notifications@github.com>
wrote:

> Users who don't install: I think we can leave this for now. One way I have
> dealt with it in the past is to have a program write an entry in the
> registry on first execution if it doesn't already find one there. That
> brings in some issues of permissions, so let's wait on this one.
>
> Agreed
>
Yes!

>  "I would prefer that newer installs of the engine overwrite older
> installs, but that I can install multiple versions of the framework." Can
> you explain that a bit more? If old ones are overwritten, then how can you
> install multiple versions? Should we have a key like v3.0, v3.2, etc?
>
> If we say that engines will continue to be backwards compatible for the
> foreseeable future, then I don't see a need to have multiple versions of
> the engine installed. If we are not sure that will be the case, then we
> could say that minor revisions (third digit) overwrite themselves, but
> major (first or second digit) install side by side. We could also make that
> cut when we introduce potentially breaking changes. Either way, it might
> make sense to include the version in the registry key. *I will make that
> change and we can decide on the install.* Sound good?
>

Yes! And fwiw, I am imagining that engines will be backward compatible and
our own _usage_ (in our runners) of the engine will also be backward
compatible. I imagine we will always use the latest engine. I think the
side-by-side install is for the benefit of 3rd party runners who may find
ways to use the engine that we don't expect.


>  Private copy: I was suggesting something a bit different. The presence
> of an engine assembly would mean privateCopy, without ever needing to
> specify it in the call. The search order would be. If a private copy is
> found meeting the min and max requirements, use it. Otherwise continue the
> search. IOW, I think we are specifying we want a private copy twice: once
> in the call and once by virtue of the fact that it's present.
>
> If we allow the privateCopy value in the call, then I am suspecting that
> all 3rd party runners would use it, depriving users of the ability to
> upgrade. It also encourages them to modify the engine for their own
> purposes, relying on special features they add in a proprietary manner.
>
> I would disagree with the presence of the engine meaning private copy. If
> I was a third party runner, I would include the engine with my install in
> case the user doesn't have it installed, but if they have a newer engine
> installed, I would use that (privateCopy = false).
>
> privateCopy = true would then be available for people that write runners
> but do not want to support using newer runners. I think it should be their
> choice, not their users because it is them, not us, that will need to
> support their software.
>

OK, you convinced me.


>  User wants to upgrade: This should probably be first, because it drives
> a lot of the rest for me. Because of past problems with TD.Net and R#, I
> wanted to give users the choice of forcing the 3rd party runner to use a
> new engine we have released. With the current API, they only have that
> choice if the vendor has allowed it. I'll allow that this might be OK. If
> the vendor isn't sufficiently responsive in terms of using a new engine,
> then pressure will be brought on them to change how they use the API and
> allow users to do that.
>
> Still, it seems to me that vendors knowing their code may be used with any
> future build of the engine would force them to write it in a safe way.
>
> Note that those who use the framework directly will already have to work
> this way. It's only possible for the framework API to use the framework
> that the user has referenced.
>
> I see your point, but I still stand by my previous points. I think we
> should encourage, but not force, developers that write runners to use the
> API correctly and to set privateCopy = false. If we do remove the
> privateCopy, I expect almost all runners will ship the engine with their
> runner effectively making it private copy. In that case, they will have the
> same incentive to use the API incorrectly. In both cases, if the user wants
> to update the engine, they can drop it in the runner directory and hope
> that it works.
>
> I am not strongly attached to this though, just making my point, so make
> the call @charliepoole <https://github.com/charliepoole> and I will
> implement it accordingly.
>

No, you're right. The pressure on vendors to allow user upgrade will have
to come from their users.


>  New point: If most vendors end up not using the engine, then maybe we
> should reduce the API to those features we want to use. I think that means
> minimum version only and no private copy.
>
> I also wonder if that will be the case, but I think we should make it as
> flexible as possible to encourage them to use it. I will leave privateCopy
> up to you as per above, but I think we can probably remove maxVersion
> since we will never know at what point compatibility will be broken and
> once that happens, you would likely update the runner and set minVersion
> instead.
>

Fair enough, let's remove maxversion but keep the rest.

Charlie

**Comment by rprouse** [comment]
I have merged in latest from master and fixed the conflicts. I have also made the two final changes that we agreed upon, removed maxVersion and check for multiple engine versions in the registry. The installer has also been updated to version the engine in the registry.

The initial PR info has been updated to reflect the changes. This should be ready to merge once it is reviewed and CI finishes running.

**Comment by rprouse** [comment]
Test failure was the random async failure that we are seeing sometimes on Travis. Unrelated to these changes.

**Comment by CharliePoole** [comment]
This looks good to me and matches what we discussed.
